File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/component/JwtProvider.java
package com.windea.demo.cloudcollect.core.component;

import io.jsonwebtoken.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.util.Date;
import java.util.Objects;

/**
 * Jwt提供器。
 */
@Component
public class JwtProvider {
	private static final Logger logger = LoggerFactory.getLogger(JwtProvider.class);

	@Value("${com.windea.security.jwtSecret}")
	private String jwtSecret;
	@Value("${com.windea.security.jwtExpiration}")
	private int jwtExpiration;


	/**
	 * 生成令牌。
	 */
	public String generateToken(Authentication authentication) {
		return Jwts.builder()
			.setSubject(((UserDetails) authentication.getPrincipal()).getUsername())
			.setIssuedAt(new Date())
			.setExpiration(generateExpiration())
			.signWith(SignatureAlgorithm.HS512, jwtSecret)
			.compact();
	}

	/**
	 * 刷新令牌。
	 */
	public String refreshToken(String token) {
		if(!isTokenNotExpired(token)) {
			return token;
		}
		return Jwts.builder()
			.setClaims(getClaims(token))
			.setIssuedAt(new Date())
			.setExpiration(generateExpiration())
			.signWith(SignatureAlgorithm.HS512, jwtSecret)
			.compact();
	}

	/**
	 * 生成过期时间。
	 */
	private Date generateExpiration() {
		return new Date(System.currentTimeMillis() + jwtExpiration * 1000);
	}

	/**
	 * 得到要求。
	 */
	public Claims getClaims(String token) {
		try {
			return Jwts.parser().setSigningKey(jwtSecret).parseClaimsJws(token).getBody();
		} catch(Exception e) {
			logger.error("Invalid Jwt format.");
			return Jwts.claims();
		}
	}

	/**
	 * 得到用户名。
	 */
	public String getUsername(String token) {
		return getClaims(token).getSubject();
	}

	/**
	 * 验证令牌。
	 */
	public boolean validateToken(String token, UserDetails userDetails) {
		return isUsernameValid(token, userDetails) && isTokenNotExpired(token);
	}

	/**
	 * 判断用户名是否合法。
	 */
	private boolean isUsernameValid(String token, UserDetails userDetails) {
		var username = getUsername(token);
		return username != null && Objects.equals(getUsername(token), userDetails.getUsername());
	}

	/**
	 * 判断令牌是否已过期。
	 */
	private boolean isTokenNotExpired(String token) {
		var expiration = getClaims(token).getExpiration();
		return expiration != null && expiration.after(new Date());
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/component/JwtProvider.kt
package com.windea.demo.cloudcollect.core.component

import com.windea.demo.cloudcollect.core.properties.*
import com.windea.demo.cloudcollect.core.service.impl.*
import io.jsonwebtoken.*
import org.springframework.security.core.userdetails.*
import org.springframework.stereotype.*
import java.util.*
import javax.servlet.http.*

@Component
class JwtProvider(
	private val userDetailsService: UserDetailsServiceImpl,
	private val jwtProperties: JwtProperties
) {
	/**从http请求中得到令牌。*/
	fun getToken(request: HttpServletRequest): String? {
		val header = request.getHeader(jwtProperties.tokenHeader)
		return if(header != null && header.startsWith(jwtProperties.tokenHead, true)) {
			header.replace(jwtProperties.tokenHead, "").trim().takeIf { it.isNotEmpty() }
		} else null
	}
	
	/**生成令牌。*/
	fun generateToken(username: String): String {
		return Jwts.builder()
			.setSubject(username)
			.setIssuedAt(Date())
			.setExpiration(generateExpiration())
			.signWith(SignatureAlgorithm.HS512, jwtProperties.secret)
			.compact()
	}
	
	/**刷新令牌。*/
	fun refreshToken(token: String): String {
		return Jwts.builder()
			.setClaims(getClaims(token))
			.setIssuedAt(Date())
			.setExpiration(generateExpiration())
			.signWith(SignatureAlgorithm.HS512, jwtProperties.secret)
			.compact()
	}
	
	private fun generateExpiration(): Date {
		return Date(System.currentTimeMillis() + jwtProperties.expiration * 1000)
	}
	
	/**得到要求。*/
	private fun getClaims(token: String): Claims {
		return try {
			Jwts.parser().setSigningKey(jwtProperties.secret).parseClaimsJws(token).body
		} catch(e: Exception) {
			println("Invalid Jwt format.")
			Jwts.claims()
		}
	}
	
	/**验证令牌。*/
	fun validateToken(token: String): UserDetails? {
		val claims = getClaims(token)
		
		//判断令牌是否过期
		val expiration = claims.expiration ?: return null
		if(expiration < Date()) return null
		
		//判断用户是否合法
		val username = claims.subject ?: return null
		return userDetailsService.loadUserByUsername(username)
	}
}
